File: Plugins/CodeEditor/Code Editors/LanguageForm.vb
Imports System
Imports System.ComponentModel
Imports System.Windows.Forms
Imports Alsing.SourceCode

''' <summary>
''' Summary description for LanguageForm.
''' </summary>
Public Class LanguageForm
    Inherits Form

    WithEvents btnCancel As Button
    WithEvents btnOK As Button
    WithEvents components As IContainer
    Public EditForm As editForm
    WithEvents imlIcons As ImageList
    WithEvents lvFileTypes As ListView
    WithEvents trvFileTypes As TreeView

    Public Sub New()
        '
        ' Required for Windows Form Designer support
        '
        InitializeComponent()
        '
        ' TODO: Add any constructor code after InitializeComponent call
        '
    End Sub

    Public Sub New(ByVal LangList As CodeEditor.SyntaxDefinitionList)

        InitializeComponent()
        trvFileTypes.Nodes.Clear()
        For Each syntax As SyntaxDefinition In LangList.GetSyntaxDefinitions()
            Dim tn As TreeNode = trvFileTypes.Nodes.Add(syntax.Name)
            tn.Tag = syntax
        Next
        trvFileTypes.SelectedNode = trvFileTypes.Nodes(0)
    End Sub

    ''' <summary>
    ''' Clean up any resources being used.
    ''' </summary>
    Protected Overrides Sub Dispose(ByVal disposing As Boolean)

        If (disposing) Then
            If (components IsNot Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub

    Private Sub LanguageForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
    End Sub

    Private Sub btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnCancel.Click
        DialogResult = DialogResult.Cancel
        Close()
    End Sub

    Private Sub btnOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnOK.Click
        OK()
    End Sub

    Private Sub trvFileTypes_AfterSelect(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles trvFileTypes.AfterSelect

        Dim syntax = CType(e.Node.Tag, SyntaxDefinition)
        lvFileTypes.Items.Clear()
        For Each ft In syntax.FileTypes
            Dim lvi As ListViewItem = lvFileTypes.Items.Add(String.Format("{0}   ({1})", ft.Name, ft.Extension))
            lvi.Tag = ft
            lvi.ImageIndex = 0
        Next
    End Sub

    Private Sub OK()
        If (lvFileTypes.SelectedItems.Count = 0) Then
            lvFileTypes.Items(0).Selected = True
        End If

        Dim syntax As SyntaxDefinition = CType(trvFileTypes.SelectedNode.Tag, SyntaxDefinition)
        Dim ft = CType(lvFileTypes.SelectedItems(0).Tag, FileType)
        Dim doc = New Document("", "", "")
        doc.Title = ("Untitled" + ft.Extension)
        doc.Path = NotepadX.Main.DefaultSaveLocation & "\"
        Dim DocName As String = String.Empty
        While String.IsNullOrEmpty(DocName)
            DocName = InputBox("Document Name (cannot be empty)")
            Dim found As Boolean = False
            For Each illegalChar In IO.Path.GetInvalidPathChars
                If DocName.Contains(illegalChar) Then found = True
            Next
            If DocName.Contains("\") Or DocName.Contains("/") _
                Or DocName.Contains(":") Then found = True
            If found Then
                MsgBox("Document Name contains illegal character!")
                DocName = ""
            End If
        End While
        doc.Document = ""
        doc.Path = doc.Path & DocName & ft.Extension
        doc.Title = doc.Path
        Dim ef As New EditForm(doc.Title, doc.Path, doc.Document, syntax)
        EditForm = ef
        DialogResult = DialogResult.OK
        Me.Close()
    End Sub

    Private Sub lvFileTypes_DoubleClick(ByVal sender As Object, ByVal e As EventArgs) Handles lvFileTypes.DoubleClick
        OK()
    End Sub

#Region " Windows Form Designer generated code "

    ''' <summary>
    ''' Required method for Designer support - do not modify
    ''' the contents of me method with the code editor.
    ''' </summary>
    Private Sub InitializeComponent()
        Me.components = New System.ComponentModel.Container()
        Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(LanguageForm))
        Me.lvFileTypes = New System.Windows.Forms.ListView()
        Me.imlIcons = New System.Windows.Forms.ImageList(Me.components)
        Me.btnOK = New System.Windows.Forms.Button()
        Me.btnCancel = New System.Windows.Forms.Button()
        Me.trvFileTypes = New System.Windows.Forms.TreeView()
        Me.SuspendLayout()
        '
        'lvFileTypes
        '
        Me.lvFileTypes.HideSelection = False
        Me.lvFileTypes.LargeImageList = Me.imlIcons
        Me.lvFileTypes.Location = New System.Drawing.Point(176, 8)
        Me.lvFileTypes.Name = "lvFileTypes"
        Me.lvFileTypes.Size = New System.Drawing.Size(272, 345)
        Me.lvFileTypes.SmallImageList = Me.imlIcons
        Me.lvFileTypes.TabIndex = 0
        Me.lvFileTypes.UseCompatibleStateImageBehavior = False
        '
        'imlIcons
        '
        Me.imlIcons.ImageStream = CType(resources.GetObject("imlIcons.ImageStream"), System.Windows.Forms.ImageListStreamer)
        Me.imlIcons.TransparentColor = System.Drawing.Color.Transparent
        Me.imlIcons.Images.SetKeyName(0, "")
        '
        'btnOK
        '
        Me.btnOK.Location = New System.Drawing.Point(285, 359)
        Me.btnOK.Name = "btnOK"
        Me.btnOK.Size = New System.Drawing.Size(75, 23)
        Me.btnOK.TabIndex = 1
        Me.btnOK.Text = "OK"
        '
        'btnCancel
        '
        Me.btnCancel.DialogResult = System.Windows.Forms.DialogResult.Cancel
        Me.btnCancel.Location = New System.Drawing.Point(373, 359)
        Me.btnCancel.Name = "btnCancel"
        Me.btnCancel.Size = New System.Drawing.Size(75, 23)
        Me.btnCancel.TabIndex = 2
        Me.btnCancel.Text = "Cancel"
        '
        'trvFileTypes
        '
        Me.trvFileTypes.HideSelection = False
        Me.trvFileTypes.Location = New System.Drawing.Point(8, 8)
        Me.trvFileTypes.Name = "trvFileTypes"
        Me.trvFileTypes.Size = New System.Drawing.Size(168, 345)
        Me.trvFileTypes.TabIndex = 3
        '
        'LanguageForm
        '
        Me.AcceptButton = Me.btnOK
        Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
        Me.CancelButton = Me.btnCancel
        Me.ClientSize = New System.Drawing.Size(458, 394)
        Me.ControlBox = False
        Me.Controls.Add(Me.trvFileTypes)
        Me.Controls.Add(Me.btnCancel)
        Me.Controls.Add(Me.btnOK)
        Me.Controls.Add(Me.lvFileTypes)
        Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog
        Me.Name = "LanguageForm"
        Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen
        Me.Text = "Select a Language"
        Me.ResumeLayout(False)

    End Sub

#End Region

End Class

File: Plugins/CodeEditor/Code Editors/LanguageForm.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using System.ComponentModel;
using System.Windows.Forms;
using Alsing.SourceCode;
namespace CodeEditor
{

    /// <summary>
    /// Summary description for LanguageForm.
    /// </summary>
    public class LanguageForm : Form
    {
        
        private Button withEventsField_btnCancel;
        public Button btnCancel {
            get { return withEventsField_btnCancel; }
            set {
                if (withEventsField_btnCancel != null) {
                    withEventsField_btnCancel.Click -= btnCancel_Click;
                }
                withEventsField_btnCancel = value;
                if (withEventsField_btnCancel != null) {
                    withEventsField_btnCancel.Click += btnCancel_Click;
                }
            }
        }
        private Button withEventsField_btnOK;
        public Button btnOK {
            get { return withEventsField_btnOK; }
            set {
                if (withEventsField_btnOK != null) {
                    withEventsField_btnOK.Click -= btnOK_Click;
                }
                withEventsField_btnOK = value;
                if (withEventsField_btnOK != null) {
                    withEventsField_btnOK.Click += btnOK_Click;
                }
            }
        }
        IContainer components;
        public EditForm EditForm;
        ImageList imlIcons;
        private ListView withEventsField_lvFileTypes;
        public ListView lvFileTypes {
            get { return withEventsField_lvFileTypes; }
            set {
                if (withEventsField_lvFileTypes != null) {
                    withEventsField_lvFileTypes.DoubleClick -= lvFileTypes_DoubleClick;
                }
                withEventsField_lvFileTypes = value;
                if (withEventsField_lvFileTypes != null) {
                    withEventsField_lvFileTypes.DoubleClick += lvFileTypes_DoubleClick;
                }
            }
        }
        private TreeView withEventsField_trvFileTypes;
        public TreeView trvFileTypes {
            get { return withEventsField_trvFileTypes; }
            set {
                if (withEventsField_trvFileTypes != null) {
                    withEventsField_trvFileTypes.AfterSelect -= trvFileTypes_AfterSelect;
                }
                withEventsField_trvFileTypes = value;
                if (withEventsField_trvFileTypes != null) {
                    withEventsField_trvFileTypes.AfterSelect += trvFileTypes_AfterSelect;
                }
            }

        }
        public LanguageForm()
        {
            Load += LanguageForm_Load;
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
            //
            // TODO: Add any constructor code after InitializeComponent call
            //
        }


        public LanguageForm(CodeEditor.SyntaxDefinitionList LangList)
        {
            Load += LanguageForm_Load;
            InitializeComponent();
            trvFileTypes.Nodes.Clear();
            foreach (SyntaxDefinition syntax in LangList.GetSyntaxDefinitions()) {
                TreeNode tn = trvFileTypes.Nodes.Add(syntax.Name);
                tn.Tag = syntax;
            }
            trvFileTypes.SelectedNode = trvFileTypes.Nodes[0];
        }

        protected override void Dispose(bool disposing)
        {
            if ((disposing)) {
                if ((components != null)) {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        private void LanguageForm_Load(object sender, EventArgs e)
        {
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
            Close();
        }

        private void btnOK_Click(object sender, EventArgs e)
        {
            OK();
        }


        private void trvFileTypes_AfterSelect(object sender, TreeViewEventArgs e)
        {
            var syntax = (SyntaxDefinition)e.Node.Tag;
            lvFileTypes.Items.Clear();
            foreach (var ft_loopVariable in syntax.FileTypes) {
                var ft = ft_loopVariable;
                ListViewItem lvi = lvFileTypes.Items.Add(string.Format("{0}   ({1})", ft.Name, ft.Extension));
                lvi.Tag = ft;
                lvi.ImageIndex = 0;
            }
        }

        private void OK()
        {
            if ((lvFileTypes.SelectedItems.Count == 0)) {
                lvFileTypes.Items[0].Selected = true;
            }

            SyntaxDefinition syntax = (SyntaxDefinition)trvFileTypes.SelectedNode.Tag;
            var ft = (FileType)lvFileTypes.SelectedItems[0].Tag;
            var doc = new Document("", "", "");
            doc.Title = ("Untitled" + ft.Extension);
            doc.Path = Microsoft.VisualBasic.FileIO.SpecialDirectories.MyDocuments + "\\";
            string DocName = "Document";
            doc.DocumentText = "";
            doc.Path = doc.Path + DocName + ft.Extension;
            doc.Title = doc.Path;
            EditForm ef = new EditForm(doc.Title, doc.Path, doc.DocumentText, syntax, ft);
            EditForm = ef;
            DialogResult = DialogResult.OK;
            this.Close();
        }

        private void lvFileTypes_DoubleClick(object sender, EventArgs e)
        {
            OK();
        }

        #region " Windows Form Designer generated code "

        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of me method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.components = new System.ComponentModel.Container();
            System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(LanguageForm));
            this.lvFileTypes = new System.Windows.Forms.ListView();
            this.imlIcons = new System.Windows.Forms.ImageList(this.components);
            this.btnOK = new System.Windows.Forms.Button();
            this.btnCancel = new System.Windows.Forms.Button();
            this.trvFileTypes = new System.Windows.Forms.TreeView();
            this.SuspendLayout();
            //
            //lvFileTypes
            //
            this.lvFileTypes.HideSelection = false;
            this.lvFileTypes.LargeImageList = this.imlIcons;
            this.lvFileTypes.Location = new System.Drawing.Point(176, 8);
            this.lvFileTypes.Name = "lvFileTypes";
            this.lvFileTypes.Size = new System.Drawing.Size(272, 345);
            this.lvFileTypes.SmallImageList = this.imlIcons;
            this.lvFileTypes.TabIndex = 0;
            this.lvFileTypes.UseCompatibleStateImageBehavior = false;
            //
            //imlIcons
            //
            this.imlIcons.ImageStream = (System.Windows.Forms.ImageListStreamer)resources.GetObject("imlIcons.ImageStream");
            this.imlIcons.TransparentColor = System.Drawing.Color.Transparent;
            this.imlIcons.Images.SetKeyName(0, "");
            //
            //btnOK
            //
            this.btnOK.Location = new System.Drawing.Point(285, 359);
            this.btnOK.Name = "btnOK";
            this.btnOK.Size = new System.Drawing.Size(75, 23);
            this.btnOK.TabIndex = 1;
            this.btnOK.Text = "OK";
            //
            //btnCancel
            //
            this.btnCancel.DialogResult = System.Windows.Forms.DialogResult.Cancel;
            this.btnCancel.Location = new System.Drawing.Point(373, 359);
            this.btnCancel.Name = "btnCancel";
            this.btnCancel.Size = new System.Drawing.Size(75, 23);
            this.btnCancel.TabIndex = 2;
            this.btnCancel.Text = "Cancel";
            //
            //trvFileTypes
            //
            this.trvFileTypes.HideSelection = false;
            this.trvFileTypes.Location = new System.Drawing.Point(8, 8);
            this.trvFileTypes.Name = "trvFileTypes";
            this.trvFileTypes.Size = new System.Drawing.Size(168, 345);
            this.trvFileTypes.TabIndex = 3;
            //
            //LanguageForm
            //
            this.AcceptButton = this.btnOK;
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.CancelButton = this.btnCancel;
            this.ClientSize = new System.Drawing.Size(458, 394);
            this.ControlBox = false;
            this.Controls.Add(this.trvFileTypes);
            this.Controls.Add(this.btnCancel);
            this.Controls.Add(this.btnOK);
            this.Controls.Add(this.lvFileTypes);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
            this.Name = "LanguageForm";
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen;
            this.Text = "Select a Language";
            this.ResumeLayout(false);

        }

        #endregion

    }
}
